magic-nix-cache/src/flakehub.rs

```diff
@@ -160,14 +179,120 @@ pub async fn init_cache(
 
     Ok(())
 }
+
+/// Refresh the GitHub Actions JWT every 2 minutes (slightly less than half of the default validity
+/// period) to ensure pushing / pulling doesn't stop working.
+async fn refresh_github_actions_jwt_worker(
+    netrc_path: std::path::PathBuf,
+    mut github_jwt: String,
+    flakehub_cache_server_clone: String,
+    api: Arc<RwLock<ApiClient>>,
+) -> Result<()> {
+    // TODO(cole-h): this should probably be half of the token's lifetime ((exp - iat) / 2), but
+    // getting this is nontrivial so I'm not going to do it until GitHub changes the lifetime and
+    // breaks this.
+    let next_refresh = std::time::Duration::from_secs(2 * 60);
+
+    // NOTE(cole-h): https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers#requesting-the-jwt-using-environment-variables
+    let mut headers = reqwest::header::HeaderMap::new();
+    headers.insert(
+        reqwest::header::ACCEPT,
+        HeaderValue::from_static("application/json;api-version=2.0"),
+    );
+    headers.insert(
+        reqwest::header::CONTENT_TYPE,
+        HeaderValue::from_static("application/json"),
+    );
+
+    let github_client = reqwest::Client::builder()
+        .user_agent(USER_AGENT)
+        .default_headers(headers)
+        .build()?;
+
+    loop {
+        match rewrite_github_actions_token(&github_client, &netrc_path, &github_jwt).await {
+            Ok(new_github_jwt) => {
+                github_jwt = new_github_jwt;
+
+                let server_config = ServerConfig {
+                    endpoint: flakehub_cache_server_clone.clone(),
+                    token: Some(attic_client::config::ServerTokenConfig::Raw {
+                        token: github_jwt.clone(),
+                    }),
+                };
+                let new_api = ApiClient::from_server_config(server_config)?;
+
+                {
+                    let mut api_client = api.write().await;
+                    *api_client = new_api;
+                }
+
+                tracing::debug!(
+                    "Stored new token in netrc and API client, sleeping for {next_refresh:?}"
+                );
+                tokio::time::sleep(next_refresh).await;
+            }
+            Err(e) => {
+                tracing::error!(
+                    ?e,
+                    "Failed to get a new JWT from GitHub, trying again in 10 seconds"
+                );
+                tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+            }
+        }
+    }
+}
+
+async fn rewrite_github_actions_token(
+    client: &reqwest::Client,
+    netrc_path: &Path,
+    old_github_jwt: &str,
+) -> Result<String> {
+    // NOTE(cole-h): https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers#requesting-the-jwt-using-environment-variables
+    let runtime_token = std::env::var("ACTIONS_RUNTIME_TOKEN")
```

Comment:
ooooh, weird:
https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect

the docs sorta imply that this should be used: `ACTIONS_ID_TOKEN_REQUEST_TOKEN` ?

but then, in another place in the docs, it mentions `ACTIONS_RUNTIME_TOKEN`?

also, I'm seeing that the latter (and presumably the former) is only valid for 6 hours?

Reply:
seems like the longer env var is what toolkit uses.
https://github.com/actions/toolkit/blob/5a736647a123ecf8582376bdaee833fbae5b3847/packages/core/src/oidc-utils.ts#L29C1-L29C4

also, googling the shorter `ACTIONS_RUNTIME_TOKEN` doesn't have a ton of hits. i'm wondering if that was deprecated in favor of the `ACTIONS_ID_TOKEN_REQUEST_TOKEN`.

gonna look a bit more.